### DISCUSSION TITLE
Color grading

### DISCUSSION THREAD

**Discussion by unicomp21** [google/filament#5193] [Q&A] [answered]
Starting to realize I don't understand how color grading works.  Could someone enlighten me or post a newbie link?

The reason for asking, it's the only place I see which is leveraging sampler_3d.  What I want is to understand how to create a 3D texture (my intension is to generate tiled 3D texture on cpu), mip map it, and use it the same way as a 2D texture/material for my meshes.  ie where the shaders start using uvw instead of just uv.  I guess this means I would have to tweak all the filament shaders as well?  Or do all the filament shaders support 3D textures?  And I'm just missing something?  My target is webgl2 or webgpu.

**Comment by romainguy** [comment]
3D textures are supported. I'm not sure what you want to know exactly?

**Comment by unicomp21** [comment, marked as answer]
Is there a webgl sample/example, like suzanne?  Which shows how to use them?

**Reply by romainguy** [reply]
There isn't but it's not terribly different from a regular 2D texture. Declare the material parameter as a sampler3D (example: https://github.com/google/filament/blob/764c695e4ebaa4f3ab70220f189e1cbe1d79dfce/filament/src/materials/colorGrading/colorGrading.mat#L10 ) and when you call the texture() function in the shader make sure to pass a vec3 as the texture coordinates argument.

**Comment by unicomp21** [comment]
Thanks!
